### DISCUSSION TITLE
Kconfig vs cmake (porting existing libraries)

### DISCUSSION THREAD

**Discussion by xR3b0rn** [zephyrproject-rtos/zephyr#80197] [Q&A]
Currently, I'm porting an existing library to zephyr. The project I'm porting is cmake based and of course uses cmake as cmake should be used.

My first naive approach of proting the library was, to go the cmake way of doing stuff. I.e. I went over the CMakeLists.txt in the root folder and worked with `if architecture == zephyr then`.

This way I was able to fully port the library. Now I'm able to use the library and fully configure the library by setting cmake options before fetching the library from github by using `FetchContent`. However, as it seems, this is not how things are done in zephyr. Zephyr requires a modules and Kconfig file in a folder called `zephyr/` and doesn't plan to fetch external modules using `FetchContent`.

From here, I don't understand why and how I should proceed...
1. I'm able to fully compile and configure the library for zephyr
2. The library itself comes with support for windows, linux and is configure for both of the OS' using the cmake-`configure`-command. My current understanding of `Kconfig` tells me, I would have to use macros generated from `Kconfig` to configure the library instead of relying on the `config.h`-file generated by the cmake-`configure`-command.
3. Should I proceed by somehow map options defined in the output of `Kconfig` to the `config.h.in`-file of the library?

From my point of view, zephyr is limiting the user of cmake too hard. I can't see any reason why I shouldn't be able to pull the library using `FetchContent` or not relying on `configure` instead of the output of `Kconfig`.

Is there any discussion on this topic?

**Comment by yashi** [comment]
Sometime a limitaiton is a good thing because without limitation (or a standard way of doing things), the Zephyr build system can't integrate external libraries or it must be too complex to be maintained.  But I can feel your pain of learning new things even though you are fluent of using plain CMake.

`FetchContent`, or a way to download external library, in Zephyr is handled by [West, Zephyr's meta-tool](https://docs.zephyrproject.org/latest/develop/west/index.html).  Thus you don't need to use `FetchContent`. Plus with West, or `west update`, you can maintain the version of your library with [`west.yml`](https://docs.zephyrproject.org/latest/develop/west/manifest.html).

Regarding Kconfig option, you can directly check in your `CMakeLists.txt` just like:

```cmake
if(CONFIG_YOUR_LIBRARY_FEATURE)
    :
endif()
```
Zephyr build system handles Kconfig and options for you.

If Zephyr build system alows user to use `configure_file`, users must check `config.h` *and* Kconfig.  Kconfig is the central repository of all configuration in Zephyr and all modules.  Your library become a port of modules.  Users of your library will launch `west build -t menuconfig` and find *all* configuration your library allows along side with other configrations. No manual checking  and searching the files for each modules.  That'd be chaos considering the number of modules we already support.

https://docs.zephyrproject.org/latest/build/kconfig/menuconfig.html

Check [picolibc](https://github.com/picolibc/picolibc/tree/main/zephyr) or any other libraries Zephyr already uses and learn how they handle Kconfig and CMake.

**Comment by xR3b0rn** [comment]
Thanks for the example of picolibc. The way picolibc handles zephyr is exactly what I came up with too. However, I also found some libraries doing it differently, what tells a lot about the unclearness of how to integrate existing libraries into zephyr and is very confusing when learning zephyr.
